### PR TITLE
Feature/47 change to django fsm

### PR DESCRIPTION
**Pull request by todd-dembrey**
Change the state management to use django FSM fields,

add some metaclass magic to make the transitions available on the submission.

### CODE REVIEW

File: opentech/apply/funds/forms.py [hunk 1]
```diff
@@ -16,14 +17,21 @@ class Meta:
...
+        choices = [(name, action) for name, action in self.instance.phase.transitions.items()]
         action_field = self.fields['action']
         action_field.choices = choices
         self.should_show = bool(choices)
 
+    def clean_action(self):
+        action = self.cleaned_data['action']
+        transition = self.instance.get_transition(action)
+        if not can_proceed(transition):
+            raise forms.ValidationError('You do not have permission to do that')
```
Comment (zerolab) [inline opentech/apply/funds/forms.py]: Can this validation error include more context. i.e. what transition is this?

File: opentech/apply/funds/models.py [hunk 2]
```diff
@@ -39,12 +40,12 @@
 from .admin_forms import WorkflowFormAdminForm
 from .blocks import CustomFormFieldsBlock, MustIncludeFieldBlock, REQUIRED_BLOCK_NAMES
 from .edit_handlers import FilteredFieldPanel, ReadOnlyPanel, ReadOnlyInlinePanel
-from .workflow import SingleStage, DoubleStage, active_statuses, get_review_statuses, review_statuses
+from .workflow import SingleStage, DoubleStage, active_statuses, get_review_statuses, review_statuses, INITAL_STATE, get_stages
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: `s/INITAL_STATE/INITIAL_STATE/g`

File: opentech/apply/funds/models.py [hunk 3]
```diff
@@ -39,12 +40,12 @@
...
 
 
 WORKFLOW_CLASS = {
-    SingleStage.name: SingleStage,
-    DoubleStage.name: DoubleStage,
+    'Request': SingleStage,
+    'Concept & Proposal': DoubleStage,
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: Can the workflow name be defined in the workflow.py? The way I see it, it still is part of workflow metadata and should be kept together with the other definitions.

File: opentech/apply/funds/models.py [hunk 4]
```diff
@@ -513,7 +507,36 @@ def current(self):
...
+                    # Attach to new class
+                    method_name = '_'.join([transition_prefix, transition_name, str(data.step)])
+                    attrs[method_name] = transition_func
+
+        def get_transition(self, transition):
+            return getattr(self, '_'.join([transition_prefix, transition, str(self.phase.step)]))
+
+        attrs['get_transition'] = get_transition
+
+        # attrs['restart'] = transition(attrs['status'], source='*', target=INITAL_STATE)(lambda x: None)
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: remove? or is this being used in your further work?

File: opentech/apply/funds/templates/funds/includes/status_bar.html [hunk 5]
```diff
@@ -1,27 +1,26 @@
...
-    {% for phase in phases %}
-        <div class="status-bar__item
-                    {% if phase == status %}
-                        status-bar__item--is-current
-                    {% elif phase < status %}
-                        status-bar__item--is-complete
-                    {% endif %}">
-            <span class="status-bar__tooltip"
-                {% if phase == status %}
-                    {# We want to display the status explicitly in case phase is a MultiStep (displays "Outcome" for name) #}
```
Comment (zerolab) [inline opentech/apply/funds/templates/funds/includes/status_bar.html]: I am still not convinced by showing the first value from multiple step phases.
When something is Under discussion, but needs more info it makes sense to show "Under discussion: needs more info", and Outcome reads much better than "Accepted", even if we are not there.

File: opentech/apply/funds/workflow.py [hunk 6]
```diff
@@ -1,244 +1,47 @@
...
 
-from typing import Dict, Iterable, Iterator, List, Sequence, Set, Type, Union, TYPE_CHECKING
 
-from django.utils.text import slugify
-
-if TYPE_CHECKING:
-    from opentech.apply.users.models import User  # NOQA
-
-
-"""
```
Comment (zerolab) [inline opentech/apply/funds/workflow.py]: Can you keep the documentation?

The changes do read better and are easier to follow, but context is king!

File: opentech/apply/funds/models.py [hunk 7]
```diff
@@ -537,28 +560,24 @@ class ApplicationSubmission(WorkflowHelpers, BaseStreamForm, AbstractFormSubmiss
...
 
     objects = ApplicationSubmissionQueryset.as_manager()
 
-    @property
-    def status_name(self):
-        return self.phase.name
-
     @property
     def stage(self):
         return self.phase.stage
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: getting `'NoneType' object has no attribute 'stage'` in https://github.com/OpenTechFund/opentech.fund/blob/6b7c63797916f9216bfe1f385499b94d51715e58/opentech/apply/funds/templates/funds/tables/column_reviews.html#L6

**Comment by zerolab** [comment]
Can you review `apply.review.views.get_form_for_stage()` for the workflow changes?

**Comment by todd-dembrey** [comment]
@zerolab I've updated the review form method.

File: opentech/apply/funds/models.py [hunk 8]
```diff
@@ -513,7 +493,36 @@ def current(self):
...
+                    # Attach to new class
+                    method_name = '_'.join([transition_prefix, transition_name, str(data.step)])
+                    attrs[method_name] = transition_func
+
+        def get_transition(self, transition):
+            return getattr(self, '_'.join([transition_prefix, transition, str(self.phase.step)]))
+
+        attrs['get_transition'] = get_transition
+
+        # attrs['restart'] = transition(attrs['status'], source='*', target=INITIAL_STATE)(lambda x: None)
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: remove now that we have cb66215 ?

Comment (todd-dembrey) [reply opentech/apply/funds/models.py]: cb66215 doesnt work with the old style statuses . I had written this when i realised it wouldn't work.

Comment (zerolab) [reply opentech/apply/funds/models.py]: 👍

**Comment by zerolab** [comment]
Let's get this in.
Will let you sort it out on dev

**Pull request merged**